File: backend/lib/yarnballs/state.ex
defmodule Yarnballs.State do
  @moduledoc """
  Represents the Yarnballs game state.
  """
  require Logger

  alias Yarnballs.Collider
  alias Yarnballs.CollisionSpace
  alias Yarnballs.Enemies
  alias Yarnballs.Missiles
  alias Yarnballs.PlayerShips
  alias Yarnballs.PlayerShip
  alias Yarnballs.Environment, as: Env
  alias Yarnballs.Spawner

  @type t :: %__MODULE__{
          missiles: Missiles.t(),
          enemies: Enemies.t(),
          ships: PlayerShips.t()
        }

  @enforce_keys [
    :missiles,
    :enemies,
    :ships
  ]
  defstruct @enforce_keys

  defimpl Jason.Encoder do
    def encode(state, opts) do
      state
      |> Map.take([
        :missiles,
        :enemies,
        :ships
      ])
      |> Map.put(:level, Yarnballs.State.level(state))
      |> Map.put(:score, Yarnballs.State.total_score(state))
      |> Jason.Encode.map(opts)
    end
  end

  def init() do
    %__MODULE__{
      missiles: Missiles.init(),
      enemies: Enemies.init(),
      ships: PlayerShips.init()
    }
  end

  def spawn_ship(state, id, name) do
    %{state | ships: PlayerShips.spawn(state.ships, id, name)}
  end

  def remove_ship(state, id) do
    %{state | ships: PlayerShips.remove(state.ships, id)}
  end

  def turn_ship(state, id, vel_angle) do
    %{state | ships: PlayerShips.turn(state.ships, id, vel_angle)}
  end

  def thrust_ship(state, id, vel_x, vel_y) do
    %{state | ships: PlayerShips.thrust(state.ships, id, vel_x, vel_y)}
  end

  def spawn_missile(state, shooter_id, x, y, vel_x, vel_y, dead) do
    case PlayerShips.fetch(state.ships, shooter_id) do
      # If player is alive, shoot missiles
      {:ok, %PlayerShip{destroyed_at: nil}} ->
        # TODO: remove `dead` if there is no respawn delay
        missiles = Missiles.spawn(state.missiles, shooter_id, x, y, vel_x, vel_y, dead)
        %{state | missiles: missiles}

      # If player is dead, try to respawn
      {:ok, _} ->
        ships = PlayerShips.respawn(state.ships, shooter_id)
        %{state | ships: ships}

      :error ->
        state
    end
  end

  def spawn_enemies(state) do
    {_, spawner} = level_with_spawner(state)

    %{state | enemies: Spawner.update(spawner, state.enemies)}
  end

  def level(state) do
    {lvl, _} = level_with_spawner(state)
    lvl
  end

  defp level_with_spawner(state) do
    case total_score(state) do
      x when x < 15 -> {0, Spawner.AFewBouncers}
      x when x < 30 -> {1, Spawner.AFewRocks}
      x when x < 60 -> {2, Spawner.AFewBouncersAndRocks}
      x when x < 90 -> {3, Spawner.Rocks}
      x when x < 120 -> {4, Spawner.BouncersAndRocks}
      x when x < 200 -> {5, Spawner.BiggerRocks}
      x when x < 280 -> {6, Spawner.BouncersAndBiggerRocks}
      x when x < 330 -> {7, Spawner.FasterRocks}
      x when x < 430 -> {8, Spawner.FasterRocksAndBiggerRocks}
      x when x < 1000 -> {9, Spawner.BouncersAndFasterRocksAndBiggerRocks}
      _ -> {10, Spawner.Madness}
    end
  end

  ## Leave for testing ^
  # x when x < 1 -> {0, Spawner.AFewBouncers}
  # x when x < 2 -> {1, Spawner.AFewRocks}
  # x when x < 3 -> {2, Spawner.AFewBouncersAndRocks}
  # x when x < 4 -> {3, Spawner.Rocks}
  # x when x < 5 -> {4, Spawner.BouncersAndRocks}
  # x when x < 6 -> {5, Spawner.BiggerRocks}
  # x when x < 7 -> {6, Spawner.BouncersAndBiggerRocks}
  # x when x < 8 -> {7, Spawner.FasterRocks}
  # x when x < 9 -> {8, Spawner.FasterRocksAndBiggerRocks}
  # x when x < 10 -> {9, Spawner.BouncersAndFasterRocksAndBiggerRocks}
  # _ -> {10, Spawner.Madness}

  def total_score(state) do
    PlayerShips.total_score(state.ships)
  end

  def update(state) do
    state
    |> spawn_enemies()
    |> update_entities()
    |> update_collisions()
  end

  defp update_collisions(state) do
    spaces = collision_spaces(state)

    # TODO: come up with better names for entity x entity collision variables
    {emcs, escs} =
      for {_, space} <- spaces, map_size(space) > 1, reduce: {MapSet.new(), MapSet.new()} do
        {emcs_acc, escs_acc} ->
          {
            MapSet.union(emcs_acc, MapSet.new(CollisionSpace.emc(space))),
            MapSet.union(escs_acc, MapSet.new(CollisionSpace.esc(space)))
          }
      end

    {enemies, missiles} = Enum.unzip(emcs)

    ships =
      escs
      |> Enum.reduce(
        state.ships,
        fn {e, s}, ships -> PlayerShips.collide_with(ships, s.id, e) end
      )

    scores = Enum.frequencies_by(missiles, & &1.shooter_id)

    %{
      state
      | enemies: Enemies.remove(state.enemies, Enum.map(enemies, & &1.id)),
        missiles: Missiles.remove(state.missiles, Enum.map(missiles, & &1.id)),
        ships: PlayerShips.increase_scores(ships, scores)
    }
  end

  defp collision_spaces(state) do
    hash = spatial_hash()

    enemies = Collider.spatial_hash(state.enemies.entities, hash, :enemies)
    missiles = Collider.spatial_hash(state.missiles.entities, hash, :missiles)
    ships = Collider.spatial_hash(Map.values(state.ships.entities), hash, :ships)

    enemies
    |> Map.merge(missiles, fn _key, m1, m2 -> Map.merge(m1, m2) end)
    |> Map.merge(ships, fn _key, m1, m2 -> Map.merge(m1, m2) end)
  end

  defp spatial_hash do
    [{0, Env.width(), 50}, {0, Env.height(), 50}]
  end

  defp update_entities(state) do
    %{
      state
      | missiles: Missiles.update(state.missiles),
        enemies: Enemies.update(state.enemies),
        ships: PlayerShips.update(state.ships)
    }
  end
end

File: yarnballs/native/yarnballs/src/state.rs
use crate::collision;
use crate::spawn;
use crate::{enemy::Enemies, missile, missile::Missiles, ship::Ships};
use rustler::NifStruct;
use std::collections::HashSet;

#[derive(NifStruct)]
#[module = "Yarnballs.State"]
pub struct State {
    missiles: Missiles,
    enemies: Enemies,
    ships: Ships,
}

impl State {
    pub fn init() -> Self {
        Self {
            missiles: Missiles::init(),
            enemies: Enemies::init(),
            ships: Ships::init(),
        }
    }

    pub fn total_score(&self) -> i64 {
        self.ships.total_score()
    }

    pub fn spawn_ship(&mut self, id: String, name: Option<String>) {
        self.ships.spawn(id, name);
    }

    pub fn turn_ship(&mut self, id: String, clockwise: bool) {
        self.ships.turn(id, clockwise);
    }

    pub fn thrust_ship(&mut self, id: String) {
        self.ships.thrust(id);
    }

    pub fn remove_ship(&mut self, id: &String) {
        self.ships.remove(id);
    }

    pub fn fire_missile_or_respawn(&mut self, id: String) {
        match self.ships.is_dead(&id) {
            Some(false) => match self.ships.spawn_missile(&id) {
                Some(missile) => self.missiles.add(missile),
                None => (),
            },
            Some(true) => self.ships.respawn(id.clone()),
            None => (),
        }
    }

    pub fn level(&self) -> u32 {
        self.level_with_spawner().0
    }

    pub fn next_level_score(&self) -> (u32, Option<u32>) {
        self.level_with_spawner().1
    }

    // TODO: Define a propery level struct. Stop repeating hard-coding level score goals.
    fn level_with_spawner(&self) -> (u32, (u32, Option<u32>), Vec<Box<dyn spawn::Spawner>>) {
        match self.total_score() {
            x if x < 15 => (0, (0, Some(15)), spawn::spawners::a_few_bouncers()),
            x if x < 30 => (1, (15, Some(30)), spawn::spawners::a_few_rocks()),
            x if x < 60 => (
                2,
                (30, Some(60)),
                spawn::spawners::a_few_bouncers_and_rocks(),
            ),
            x if x < 90 => (3, (60, Some(90)), spawn::spawners::rocks()),
            x if x < 120 => (4, (90, Some(120)), spawn::spawners::bouncers_and_rocks()),
            x if x < 200 => (5, (120, Some(200)), spawn::spawners::bigger_rocks()),
            x if x < 280 => (
                6,
                (200, Some(280)),
                spawn::spawners::bouncers_and_bigger_rocks(),
            ),
            x if x < 330 => (7, (280, Some(330)), spawn::spawners::faster_rocks()),
            x if x < 430 => (
                8,
                (330, Some(430)),
                spawn::spawners::faster_rocks_and_bigger_rocks(),
            ),
            x if x < 1000 => (
                9,
                (430, Some(1000)),
                spawn::spawners::bouncers_faster_rocks_and_bigger_rocks(),
            ),
            // TODO: these levels aren't supposed to be beatable
            x if x < 2000 => (10, (1000, Some(2000)), spawn::spawners::madness()),
            // TODO: this level has noticeable lag
            x if x < 5000 => (11, (2000, Some(5000)), spawn::spawners::overkill()),
            _ => (12, (5000, None), spawn::spawners::overbounce()),
        }
    }

    pub fn update(&mut self) {
        self.missiles.update();
        self.enemies.update();
        self.ships.update();

        self.update_collisions();
        self.spawn_enemies();
    }

    fn spawn_enemies(&mut self) {
        self.enemies.spawn(self.level_with_spawner().2);
    }

    pub fn update_collisions<'a>(&mut self) {
        // Initialize spatial hash with collidable entities
        let mut sh = collision::new_spatial_hash::<'a>();
        for m in &mut self.missiles.entities {
            sh.insert_missile(m);
        }
        for e in &mut self.enemies.entities() {
            sh.insert_enemy(e);
        }
        for e in &mut self.ships.entities() {
            sh.insert_ship(e);
        }

        // Apply enemy-missile and ship-enemy collisions
        let mut enemy_missile_collisions: HashSet<collision::EnemyMissileCollision> =
            HashSet::new();
        let mut ship_enemy_collisions: HashSet<collision::ShipEnemyCollision> = HashSet::new();
        sh.spaces().for_each(|s| {
            enemy_missile_collisions.extend(s.enemy_missile_collisions());
            ship_enemy_collisions.extend(s.ship_enemy_collisions());
        });
        self.apply_enemy_missile_collisions(
            enemy_missile_collisions.into_iter().collect::<Vec<_>>(),
        );
        self.apply_ship_enemy_collisions(ship_enemy_collisions.into_iter().collect::<Vec<_>>());
    }

    pub fn apply_enemy_missile_collisions(&mut self, emcs: Vec<collision::EnemyMissileCollision>) {
        // apply entity-specific collision effects
        self.missiles.apply_enemy_collisions(&emcs);
        self.enemies.apply_missile_collisions(&emcs);

        // increase scores
        let ids: Vec<missile::ID> = emcs.into_iter().map(|(_, m)| m).collect();
        self.missiles
            .shooter_counts(&ids)
            .into_iter()
            .for_each(|(id, pts)| self.ships.increase_score(id, pts));
    }

    pub fn apply_ship_enemy_collisions(&mut self, secs: Vec<collision::ShipEnemyCollision>) {
        secs.into_iter()
            .for_each(|(sid, eid)| match self.enemies.entities.get(&eid) {
                None => (),
                Some(enemy) => self.ships.collide_with(sid, enemy),
            });
    }
}
